src/hevm/src/EVM/SymExec.hs

```diff
@@ -349,33 +353,39 @@ leaves (Node _ xs) = concatMap leaves xs
 -- | Symbolically execute the VM and check all endstates against the postcondition, if available.
 -- Returns `Right (Tree BranchInfo)` if the postcondition can be violated, or
 -- or `Left (Tree BranchInfo)`, if the postcondition holds for all endstates.
-verify :: VM -> Maybe Integer -> Maybe (Fetch.BlockNumber, Text) -> Maybe Postcondition -> Query (Either (Tree BranchInfo) (Tree BranchInfo))
+verify :: VM -> Maybe Integer -> Maybe (Fetch.BlockNumber, Text) -> Maybe Postcondition -> Query VerifyResult
 verify preState maxIter rpcinfo maybepost = do
   smtState <- queryState
   tree <- doInterpret (Fetch.oracle (Just smtState) rpcinfo False) maxIter preState
   case maybepost of
     (Just post) -> do
       let livePaths = pruneDeadPaths $ leaves tree
-      -- can also do these queries individually (even concurrently!). Could save time and report multiple violations
+          -- have we hit max iterations for any path?
+          maxReached = or $ mapMaybe (flip maxIterationsReached maxIter) livePaths
+          -- is there any path which can possibly violate the postcondition?
+          -- can also do these queries individually (even concurrently!). Could save time and report multiple violations
           postC = sOr $ fmap (\postState -> (sAnd (fst <$> view constraints postState)) .&& sNot (post (preState, postState))) livePaths
-      -- is there any path which can possibly violate
-      -- the postcondition?
       resetAssertions
       constrain postC
       io $ putStrLn "checking postcondition..."
       checkSat >>= \case
         Unk -> do io $ putStrLn "postcondition query timed out"
-                  return $ Left tree
-        Unsat -> do io $ putStrLn "Q.E.D."
-                    return $ Left tree
-        Sat -> return $ Right tree
+                  return $ Timeout tree
+        Unsat -> do
+          if maxReached
+            then do
+              io $ putStrLn "WARNING: max iterations reached, execution halted prematurely"
+              io $ putStrLn "no postcondition violations discovered"
```

Comment:
we should maybe also fail here?

Reply:
yeah I wasn't quite sure what was best, I guess I feel like it's more of a warning than a hard error, so we shouldn't exit with a failure, but I'm fine to change it if you have strong opinions...